xbmc/addons/AddonInfo.cpp

```diff
@@ -151,7 +414,448 @@ CAddonInfo::CAddonInfo(std::string id, TYPE type)
     m_mainType(type),
     m_packageSize(0)
 {
+  CAddonType addonType(m_mainType, this, nullptr);
+  m_types.push_back(addonType);
+}
+
+bool CAddonInfo::IsType(TYPE type) const
+{
+  return (m_mainType == type || ProvidesSubContent(type));
+}
+
+const CAddonType* CAddonInfo::Type(TYPE type) const
+{
+  if (type == ADDON_UNKNOWN)
+    return &m_types[0];
+
+  for (auto& addonType : m_types)
+  {
+    if (addonType.Type() == type)
+      return &addonType;
+  }
+  return nullptr;
+}
+
+std::string CAddonInfo::MainLibPath() const
+{
+  if (m_types.empty())
+    return "";
+  return m_types[0].LibPath();
+}
+
+bool CAddonInfo::ProvidesSubContent(const TYPE& content, const TYPE& mainType/* = ADDON_UNKNOWN*/) const
+{
+  if (content == ADDON_UNKNOWN)
+    return false;
+
+  for (auto addonType : m_types)
+  {
+    if ((mainType == ADDON_UNKNOWN || addonType.Type() == mainType) && addonType.ProvidesSubContent(content))
+      return true;
+  }
+
+  return false;
+}
+
+bool CAddonInfo::ProvidesSeveralSubContents() const
+{
+  int contents = 0;
+  for (auto addonType : m_types)
+    contents += addonType.ProvidedSubContents();
+  return contents > 0 ? true : false;
+}
+
+std::string CAddonInfo::SerializeMetadata()
```

Comment:
The point here is having some basic separation of concern. CAddonDatabse is responsible for mapping an unmapping. By exposing the raw unmapping `metadata` value, CAddonInfo now has to details about how things are stored in the database.

I don't see you put reading/writing from the db `Dataset` in this class so why do it with `metadata`? (hint: you don't because mapping of `Dataset` is the entire job of the database). The `metadata` value is the same thing as `Dataset`, it just happens to be json right now.